subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/impl/DirectorySnapshotterTest.groovy

```diff
@@ -150,6 +154,30 @@ class DirectorySnapshotterTest extends Specification {
         ] as Set
     }
 
+    @Requires(TestPrecondition.FILE_PERMISSIONS)
+    def "unreadable files and directories are snapshotted as missing"() {
+        given:
+        def rootDir = tmpDir.createDir("root")
+        rootDir.file('readableFile').createFile()
+        rootDir.file('readableDirectory').createDir()
+        rootDir.file('unreadableFile').createFile().makeUnreadable()
+        rootDir.file('unreadableDirectory').createDir().makeUnreadable()
+
+        when:
+        def snapshot = directorySnapshotter.snapshot(rootDir.absolutePath, directoryWalkerPredicate(new PatternSet()), new AtomicBoolean(false))
+
+        then:
+        assert snapshot instanceof DirectorySnapshot
+        snapshot.children.collectEntries { [it.name, it.class] } == [
+            readableFile: RegularFileSnapshot,
+            readableDirectory: DirectorySnapshot,
+            unreadableFile: MissingFileSnapshot,
+            unreadableDirectory: MissingFileSnapshot
+        ]
+        cleanup:
+        rootDir.listFiles()*.readable = true
```

Comment:
would leave the files in an uncleanable state otherwise (i.e. `./gradlew clean` could fail)

Reply:
Why not put these comments in the code instead? I mean in general, not just this one. They seem to be useful whenever anyone reads the code, not just during this here review.

Reply:
I usually put these comments for the changes, not for the code specifically. Some may be useful for the code, too, but in those cases maybe it's a smell to refactor the code instead.
Is there any comment here you think I should add to the code instead?

Reply:
This one in particular would save some trouble for the next person unfamiliar with how deleting unreadable directories work. I think it should go in the code.

Reply:
The rest of the tests contain similar cleanups, would you like me to comment on all of them?

Reply:
See: https://github.com/gradle/gradle/commit/4f6641a620cf9e2fac6cd72c8ed979eb534d31ff

Reply:
This comment only applies where unreadable directories in particular need to be cleaned up. It would make sense to add the comment to wherever that happens.

BTW, I don't feel strongly, but it could even make sense to replace the cleanup here with something like:

```suggestion
        unreadableDirectory.setReadable(true, false)
```

Reply:
Isn't that what the `cleanup:` block means?
Also, it's not just about unreadable directories, it's about cleaning up files that we've created that could be hard to clean up otherwise:
e.g. doing a `./gradlew clean` without the block will result in:
<img width="1030" alt="Screen Shot 2019-07-04 at 17 58 14" src="https://user-images.githubusercontent.com/1841944/60678638-760e5300-9e85-11e9-86ea-484f4074c151.png">

Reply:
Added `makereadable` to `TestFile`